File: back/src/main/java/csh/back/domain/vote/user/service/VoteUserService.java
package csh.back.domain.vote.user.service;

import csh.back.domain.trip.member.entity.TripMember;
import csh.back.domain.trip.member.repository.TripMemberRepository;
import csh.back.domain.vote.item.entity.VoteItem;
import csh.back.domain.vote.user.dto.response.VoteUserSaveResponseDto;
import csh.back.domain.vote.user.entity.VoteUser;
import csh.back.domain.vote.user.repository.VoteUserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Random;

@Transactional(readOnly = true)
@RequiredArgsConstructor
@Service
public class VoteUserService {

    private final VoteUserRepository voteUserRepository;
    private final TripMemberRepository tripMemberRepository;

    private final Integer DEFAULT_UPDATE_COUNT = 0;

    @Transactional
    public VoteUserSaveResponseDto saveVoteUser(VoteItem voteItem, Long tripId, Long memberId) {

        TripMember tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripId)
                .orElseThrow(RuntimeException::new);
        VoteUser voteUser = voteUserRepository.findByVoteIdAndTripMemberId(
                voteItem.getVote().getId(), tripMember.getId())
                .orElse(null);
        if(voteUser == null) {
            voteUser = VoteUser.builder()
                    .vote(voteItem.getVote())
                    .voteItem(voteItem)
                    .tripMember(tripMember)
                    .updateCount(DEFAULT_UPDATE_COUNT)
                    .build();
            VoteUser saved = voteUserRepository.save(voteUser);
            return VoteUserSaveResponseDto.from(saved);
        }
        if(voteUser.getUpdateCount() == 2) throw new RuntimeException();
        voteUser.updateVoteItemAndincreaseUpdateCount(voteItem);
        return VoteUserSaveResponseDto.from(voteUser);
    }
}

File: back/src/main/kotlin/csh/back/domain/vote/user/service/VoteUserService.kt
package csh.back.domain.vote.user.service

import csh.back.domain.trip.member.repository.TripMemberRepository
import csh.back.domain.vote.item.entity.VoteItem
import csh.back.domain.vote.user.dto.response.VoteUserSaveResponseDto
import csh.back.domain.vote.user.entity.VoteUser
import csh.back.domain.vote.user.repository.VoteUserRepository
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class VoteUserService(
    private val voteUserRepository: VoteUserRepository,
    private val tripMemberRepository: TripMemberRepository,
) {

    @Transactional
    fun saveVoteUser(voteItem: VoteItem, tripGroupId: Long, memberId: Long): VoteUserSaveResponseDto {
        val tripMember = tripMemberRepository.findByMemberIdAndTripGroupId(memberId, tripGroupId)
            .orElseThrow(::RuntimeException)
        val voteUser = voteUserRepository
            .findByVoteIdAndTripMemberId(voteItem.vote.id!!, tripMember.id!!)
            .orElse(null)

        if (voteUser == null) {
            return try {
                val saved = voteUserRepository.saveAndFlush(
                    VoteUser(
                        vote = voteItem.vote,
                        voteItem = voteItem,
                        tripMember = tripMember,
                        updateCount = DEFAULT_UPDATE_COUNT,
                    ),
                )
                VoteUserSaveResponseDto.from(saved)
            } catch (e: DataIntegrityViolationException) {
                // 동시성으로 경합에서 진 경우 - 이긴 쪽이 만든 row를 재조회해 정상 흐름으로 이어감
                val winner = voteUserRepository.findByVoteIdAndTripMemberId(voteItem.vote.id!!, tripMember.id!!)
                    .orElseThrow(::RuntimeException)
                if (winner.updateCount != 2) winner.updateVoteItemAndincreaseUpdateCount(voteItem)
                VoteUserSaveResponseDto.from(winner)
            }
        }
        if (voteUser.updateCount == 2) throw RuntimeException()
        voteUser.updateVoteItemAndincreaseUpdateCount(voteItem)
        return VoteUserSaveResponseDto.from(voteUser)
    }

    companion object {
        private const val DEFAULT_UPDATE_COUNT = 0
    }
}